### PR TITLE
[WIP] Add support for an existing resource group

### PR DESCRIPTION
**Pull request by justaugustus**
Signed-off-by: Stephen Augustus <saugustus@vmware.com>

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note

```

### CODE REVIEW

File: pkg/cloud/azure/services/internalloadbalancers/internalloadbalancers.go [hunk 1]
```diff
@@ -157,13 +157,13 @@ func (s *Service) Delete(ctx context.Context, spec azure.Spec) error {
 		return errors.New("invalid internal load balancer specification")
 	}
 	klog.V(2).Infof("deleting internal load balancer %s", internalLBSpec.Name)
-	f, err := s.Client.Delete(ctx, s.Scope.ClusterConfig.ResourceGroup, internalLBSpec.Name)
+	f, err := s.Client.Delete(ctx, s.Scope.ResourceGroup().Name, internalLBSpec.Name)
```
Comment (soggiest) [inline pkg/cloud/azure/services/internalloadbalancers/internalloadbalancers.go]: why do we use `f` here? Is it worth it to change to more descriptive variable names as we go on?

Comment (soggiest) [reply pkg/cloud/azure/services/internalloadbalancers/internalloadbalancers.go]: I also know we use `f` elsewhere in this PR and the code, this is just where I noticed it

**Comment by awesomenix** [comment]
Whats the reasoning behind this PR? Why only resourcegroup and vnet, how about other resources? 
If the idea is to support externally managed resources, it would be better to decouple cluster object from machine actuator making it easier and simpler to do so.

Also can we create right interfaces for configuration and not modify the existing services?

Create a new config service which 

- provides configuration for any resource example resourcegroup, vnet, subnet etc. 
- Get provides the configuration, meanwhile CreateOrUpdate updates the configuration
- Delete would mostly be noop because it has no meaning.

Above would help in testing configuration service as well.

If the intention is to move to Reconcile model do not modify existing service interface, rather create another layer on top to use existing services and create a reconciler

Example:

ManagedResourceGroupReconciler() {
  Get()
  CreateOrUpdate()
}

Also please add tests for any new functionality and break this PR into multiple so its easier to review.

Ideally if you are changing a spec, this should be created in a new api version v1alpha2, since we have release 0.1.0.

File: cmd/clusterctl/examples/azure/generate-yaml.sh [hunk 2]
```diff
@@ -28,10 +28,11 @@ RANDOM_STRING=$(date | md5sum | head -c8)
 # Azure settings.
 export LOCATION="${LOCATION:-eastus2}"
 export RESOURCE_GROUP="${RESOURCE_GROUP:-capi-${RANDOM_STRING}}"
+# TODO: Remove once existing resource group logic is handled by tagging.
+export RESOURCE_GROUP_MANAGED="${RESOURCE_GROUP_MANAGED:-yes}"
 
 # Cluster settings.
 export CLUSTER_NAME="${CLUSTER_NAME:-test1}"
-export VNET_NAME="${VNET_NAME:-}"
```
Comment (awesomenix) [inline cmd/clusterctl/examples/azure/generate-yaml.sh]: PR says adding support for existing resource group, but i see changes for removing vnet as well, can you please revert these changes

File: config/crds/azureprovider_v1alpha1_azureclusterproviderspec.yaml [hunk 3]
```diff
@@ -474,7 +474,21 @@ spec:
               type: object
           type: object
         resourceGroup:
-          type: string
+          properties:
+            id:
+              description: ID is the identifier of the resource group.
+              type: string
+            managed:
```
Comment (awesomenix) [inline config/crds/azureprovider_v1alpha1_azureclusterproviderspec.yaml]: can this be a bool instead of a yes or no? ismanaged : true / ismanaged: false, default can be false.

Comment (awesomenix) [reply config/crds/azureprovider_v1alpha1_azureclusterproviderspec.yaml]: actually curious why isnt this implemented using tags? since we have services now, its really easy to add these tags in CreateOrUpdate?

File: pkg/cloud/azure/interfaces.go [hunk 4]
```diff
@@ -30,9 +28,17 @@ type Spec interface {
...
+	Reconcile(ctx context.Context, spec Spec) error
+	Delete(ctx context.Context, spec Spec) error
+}
+
+// GetterService is a temporary interface used by components which still require Get methods.
+// Once all components move to storing provider information within the relevant
+// Cluster/Machine specs, this interface should be removed.
+type GetterService interface {
 	Get(ctx context.Context, spec Spec) (interface{}, error)
-	CreateOrUpdate(ctx context.Context, spec Spec) error
```
Comment (awesomenix) [inline pkg/cloud/azure/interfaces.go]: Please do not change these interfaces, a Service always implements Get/CreateOrUpdate/Delete. A reconciler on other hand implements reconcile and delete.

We would want to define reconcilers in a common location if its reused in multiple places which isnt the case and only cluster actuator uses them. Please move the reconcile logic to the actuator themselves

File: pkg/cloud/azure/services/groups/groups.go [hunk 5]
```diff
@@ -23,28 +23,79 @@ import (
...
+	"sigs.k8s.io/cluster-api-provider-azure/pkg/cloud/azure/converters"
 )
 
+// Reconcile reconciles information about a resource.
+func (s *Service) Reconcile(ctx context.Context, _ azure.Spec) error {
+	klog.V(2).Infof("reconciling resource group %s", s.Scope.ResourceGroup().Name)
+	rg, err := s.Get(s.Scope.Context)
+	if err != nil {
+		switch {
+		case s.Scope.ResourceGroup().IsProvided():
```
Comment (awesomenix) [inline pkg/cloud/azure/services/groups/groups.go]: Create a Spec which specifies the name and Get/CreateOrUpdate/Delete just works on that spec rather than polluting this function.

That way a reconciler does a Get -> if managed does not call createorupdate, similar to reconciler delete -> does a get if managed does not call service.delete.

File: pkg/cloud/azure/services/internalloadbalancers/internalloadbalancers.go [hunk 6]
```diff
@@ -43,7 +43,7 @@ func (s *Service) Get(ctx context.Context, spec azure.Spec) (interface{}, error)
 		return network.LoadBalancer{}, errors.New("invalid internal load balancer specification")
 	}
 	//lbName := fmt.Sprintf("%s-api-internallb", s.Scope.Cluster.Name)
-	lb, err := s.Client.Get(ctx, s.Scope.ClusterConfig.ResourceGroup, internalLBSpec.Name, "")
+	lb, err := s.Client.Get(ctx, s.Scope.ResourceGroup().Name, internalLBSpec.Name, "")
```
Comment (awesomenix) [inline pkg/cloud/azure/services/internalloadbalancers/internalloadbalancers.go]: If resource group is a configuration, please move this to spec for each service using it instead of this

File: pkg/cloud/azure/services/internalloadbalancers/internalloadbalancers.go (hunk 1 above)
Comment (justaugustus) [reply pkg/cloud/azure/services/internalloadbalancers/internalloadbalancers.go]: Thanks for noticing this, @soggiest. I've updated the variables to `future`, which indicate that they are placeholders for future responses from Azure.

**Comment by justaugustus** [comment]
> Whats the reasoning behind this PR? Why only resourcegroup and vnet, how about other resources?

I've opened a GitHub issue to discuss existing resources and provided a more detailed description there: https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/178

In short, the reason to start with resource group and vnet is that the creation of those Azure resources are the catalyst for all other Azure resources that we utilize in capz.

> If the idea is to support externally managed resources, it would be better to decouple cluster object from machine actuator making it easier and simpler to do so.

The cluster object and machine actuator are intrisicly linked, by upstream Cluster API design. If there is a future plan to separate the two, it should be discussed with the larger Cluster API community well prior to any implementation.

Breaking that link is neither required, nor in scope for the change I'm proposing here.

> Also can we create right interfaces for configuration and not modify the existing services?
> 
> Create a new config service which
> 
> * provides configuration for any resource example resourcegroup, vnet, subnet etc.
> * Get provides the configuration, meanwhile CreateOrUpdate updates the configuration
> * Delete would mostly be noop because it has no meaning.
> 
> Above would help in testing configuration service as well.
> 
> If the intention is to move to Reconcile model do not modify existing service interface, rather create another layer on top to use existing services and create a reconciler
> 
> Example:
> 
> ManagedResourceGroupReconciler() {
> Get()
> CreateOrUpdate()
> }

I've added design details related to this on a previously opened GitHub issue: https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/110

To summarize here:
- I don't believe we actually are using the right abstractions currently, hence the change
- We absolutely should not be defining yet another means of configuration. We already have the means to do so via the clusters.yaml, machines.yaml, provider-components.yaml. If new functionality needs to be added, it should be extended via the API types and represented in the relevant specs, as is done in this PR.
- I think we have enough layers. I'm not sure that we should be adding yet another abstraction layer.

> Also please add tests for any new functionality and break this PR into multiple so its easier to review.

I've broken out the vnet changes into a separate PR (#177), which I'll rebase once this merges. 

> Ideally if you are changing a spec, this should be created in a new api version v1alpha2, since we have release 0.1.0.

To quote [semver](https://semver.org/#spec-item-4):
> Major version zero (0.y.z) is for initial development. Anything may change at any time. The public API should not be considered stable.

We don't currently adhere to any API contract for capz specifically.
We do provide some guarantees in our [support policy](https://github.com/kubernetes-sigs/cluster-api-provider-azure#support-policy), but those guarantees are around supported versions of Kubernetes and Cluster API.

That is to say, our API will move to v1alpha2 when CAPI does.

File: pkg/cloud/azure/services/internalloadbalancers/internalloadbalancers.go (hunk 6 above)
Comment (justaugustus) [reply pkg/cloud/azure/services/internalloadbalancers/internalloadbalancers.go]: See my comments below. We should be leveraging the defined API types as opposed to a package-level Spec.

File: cmd/clusterctl/examples/azure/generate-yaml.sh (hunk 2 above)
Comment (justaugustus) [reply cmd/clusterctl/examples/azure/generate-yaml.sh]: This variable is unused currently, hence the removal. The logic is wired up in #177 instead.

File: config/crds/azureprovider_v1alpha1_azureclusterproviderspec.yaml (hunk 3 above)
Comment (justaugustus) [reply config/crds/azureprovider_v1alpha1_azureclusterproviderspec.yaml]: We need to use string values, as there's no guarantee the various tools that manipulate the yaml will interpret `"true"|"false"` as `true|false`. I use `"yes"|"no"` here instead so that there's no confusion about bool vs string.

There's a larger discussion about tagging in cloud providers vs cluster API providers (ref: https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/704, https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/706) which suggests that there still are some design decisions that need to be made, hence my hesitation to start introducing them here.

This is a lightweight solution in the interim and can refactored easily once there's some consensus about how to tag Cluster API objects.

cc-ing @johnharris85, as he intends to work on writing the tagging bits.

File: pkg/cloud/azure/interfaces.go (hunk 4 above)
Comment (justaugustus) [reply pkg/cloud/azure/interfaces.go]: See my comments below.

File: pkg/cloud/azure/services/groups/groups.go (hunk 5 above)
Comment (justaugustus) [reply pkg/cloud/azure/services/groups/groups.go]: See my comments below and #110.

**Comment by justaugustus** [comment]
Moved the reconcile changes out to #180. We'll need to merge that one before I can wrap this PR up.

**Comment by justaugustus** [comment]
Implementation is complete here, but writing tests for the resource groups package are blocked on adding mock interfaces to the project.

Mocks are ready for review in #182.

/assign @awesomenix @soggiest

**Comment by awesomenix** [comment]
@justaugustus PR looks good overall, can you please add the tests using existing mocks, other than that PR is ready for checkin

File: pkg/cloud/azure/converters/resourcegroup.go [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2018-05-01/resources"
+	"sigs.k8s.io/cluster-api-provider-azure/pkg/apis/azureprovider/v1alpha1"
+)
+
+// SDKToVnet converts azure representation to internal representation
+func SDKToResourceGroup(rg resources.Group, managed string) *v1alpha1.ResourceGroup {
+	return &v1alpha1.ResourceGroup{
+		ID:      *rg.ID,
```
Comment (feiskyer) [inline pkg/cloud/azure/converters/resourcegroup.go]: any possible ID or name are nil?

**Comment by justaugustus** [comment]
@feiskyer -- thanks for taking a look!
There are a few things I have to clean up in the tests before they're good to go.
I should be able to finish it up this week. :)

File: pkg/cloud/azure/services/groups/groups_test.go [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			serviceMock := mocks.NewMockService(mockCtrl)
+
+			tc.expect(serviceMock.EXPECT())
+
+			err := serviceMock.Delete(context.Background(), tc.input)
```
Comment (awesomenix) [inline pkg/cloud/azure/services/groups/groups_test.go]: Shouldnt this be testing that the delete didnt actually happen?

File: pkg/cloud/azure/services/groups/groups_test.go [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			serviceMock := mocks.NewMockService(mockCtrl)
+
+			tc.expect(serviceMock.EXPECT())
+
+			err := serviceMock.Reconcile(context.Background(), tc.input)
```
Comment (awesomenix) [inline pkg/cloud/azure/services/groups/groups_test.go]: Same here shouldnt this be testing that create didnt actually happen? Would be better to test both cases here

Comment (serbrech) [reply pkg/cloud/azure/services/groups/groups_test.go]: What logic is this testing?
I was expecting to see an real implementation instantiated with fake dependencies.
Here, the system under test seems to be serviceMock, which is itself a fake/mock?

File: pkg/cloud/azure/services/groups/groups_test.go (hunk 8 above)
Comment (serbrech) [reply pkg/cloud/azure/services/groups/groups_test.go]: same as above, if you call delete on the mock, what logic is running? what are we testing?

File: pkg/cloud/azure/services/groups/groups_test.go [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+			err := serviceMock.Delete(context.Background(), tc.input)
+			if err != nil {
+				t.Fatalf("did not expect error calling a mock: %v", err)
+			}
+		})
+	}
+}
+
+/*
+func TestGet(t *testing.T) {
```
Comment (serbrech) [inline pkg/cloud/azure/services/groups/groups_test.go]: Should this be deleted?

**Comment by justaugustus** [comment]
@awesomenix @serbrech -- sorry, that's copy/pasta that I pushed up from my local. Putting this back to WIP until I have a moment to clean it up.

**Comment by juan-lee** [comment]
Support for BYORG could be implemented by just accepting an existing RG during Reconcile.

I know there is concern around cleaning up the RG, but I'd say we should make the call to never delete the RG. The main reason for this is because we can't know what else the cluster operator has put in the RG. We don't want to accidentally delete someone's SQL database or similar.

**Comment by rbitia** [comment]
@justaugustus should we close this and make a new issue? Or do you want to update this PR?

**Comment by justaugustus** [comment]
This was done on `v1alpha1` and will be a mess to rebase.
Closing in favor of @CecileRobertMichon's [future work](https://kubernetes.slack.com/archives/CEX9HENG7/p1572629244104200).
/close
